README
Readme for XML-SAX-Cacheable

XML::SAX::Cacheable extends the interface for XML::SAX to include methods for
controlling caching of output.  It's default behaviour is to allow the output
of the SAX handler to be cached indefinitely.

For more details, see the pod documentation in XML::SAX::Cacheable
For licensing, see the LICENSE file in this distribution.
To install:

  perl Makefile.PL
  make
  make test
  sudo make install

Will probably do it.

Please send any bug reports or praise to Chris Leishman <chris@leishman.org>.

Enjoy!


----------------------------------------------
Design Notes.

The SAX caching interface consists of two main compontents, the
XML::SAX::Cacheable interface and the XML::SAX::Cache::Entry interface.

XML::SAX::Cacheable is designed as an extension to the normal XML::SAX
interface, providing standard methods to facilitate the enabling of caching
and the discovery of details relevant to the cached version, such as whether
the cached version is 'up-to-date' and the presence of dependancies.  The
intention of this interface is to give the user maximum flexiblity in
determining if, and deciding when, a SAX filter's output should be cached, or
the cached version retrieved.

The XML::SAX::Cacheable methods are detailed in the pod documentation (and
that is the definitive source), however I'll give a quick run down of the
methods and their rationale.

$x->cacheable()
 This is used to indicate whether it is possible to cache the output of the
 filter at all.  If a filters output is entirely dynamic, then it should
 define this method so that it returns a false value.  This allows the user to
 avoid calling cache_enable for this filter and for any filters that obtain
 their input from this filter.

 If a filter returns a false value here (undef) then it shouldn't have to
 worry about implementing any of the other methods defined by this interface -
 and the user should ensure they never call them.

$x->cache_key()
 In some situations it is possible that the output from a filter will not be
 consistent for every run, but it is not truely dynamic either - instead being
 from a (small) set of possible inputs that will reoccur with regularity.  For
 these situations the filter can return a 'key' from this method, which will
 be used to distinguish which cached version to use.

 Typically when a filter returns a 'key' from this method, it is used to
 create a variant in the cache for this filter and for any downstream filters.
 For example, if a caching system like the Cache modules is in use, the user
 may decide on a set of Cache keys, one for each filter in a pipeline.  Then,
 after getting a 'key' back from a filter via this method, the user will
 modify the Cache key for this filter and all downstream filters by postfixing
 this 'key', thus creating variant versions.  Obviously this should be used
 with caution and each module that implements this should probably make some
 high level API available to allow the user to limit or disable this
 functionality.

$x->cache_immutable($cache_entry)
 Sometimes it is desirable to know if the output of the filter depended on
 anything other than the input, and this method can be used to determine that.
 Obviously it entirely up to the filter author how they determine this
 information.
 
 For example, given two filters in a pipeline and the first filter has
 dependancies and the second does not - it would make sense to only cache the
 output of the second filter and not bother caching the output of the first
 (instead only caching the validity object).

 Some filters will require some magic in implementing this method, which is
 discussed in the pod documentation.

$x->cache_check($cache_entry)
 Returns a boolean value indicating whether the cache is up-to-date.  How the
 filter wishes to determine this is entirely arbitrary - but it will usually
 retrieve the validity object and use that in some way.  For example, a filter
 might store a list of all file dependancies and their last modified
 timestamps in the validity object.  Then, when this method is invoked, it
 would retrieve the validity object and compare all the last modified
 timestamps to determine if it should regenerate it's output.

$x->cache_enable($cache_entry, [ $validity_only ])
 To enable caching of a filters output, this method is invoked.  The filter
 itself is responsible for setting everything up so that the output will be
 properly cached - AND so that any changes can be reverted by calling
 cache_disable.  The default implementation of this method takes care of
 everything, but some filters can possibly be more optimal in the way they
 handle this.


Changes
2006-02-06 Chris Leishman <chris@leishman.org>

 * Initial release (1.00)


t/TestFilter.pm
package TestFilter;

use strict;
use warnings;

use XML::SAX::Base;
use XML::SAX::Cacheable;
our @ISA = qw(XML::SAX::Base XML::SAX::Cacheable);

sub cache_validity {
	my $self = shift;
	return $self->{TestFilterValidity};
}

sub cache_expiry {
	my $self = shift;
	return $self->{TestFilterExpires};
}

sub test_set_validity {
	my $self = shift;
	my ($validity) = @_;
	$self->{TestFilterValidity} = $validity;
}

sub test_set_expiry {
	my $self = shift;
	my ($expires) = @_;
	$self->{TestFilterExpires} = $expires;
}

1;


lib/XML/SAX/Cache/Tee.pm
=head1 NAME

XML::SAX::Cache::Tee - Helper class for defaul XML::SAX::Cacheable

=head1 DESCRIPTION

This module is a helper for the default implementation of the
XML::SAX::Cacheable interface and is not for normal usage.
It defines a wrapper for XML::Filter::Tee that correctly sets all handlers and
also takes care of storing validity objects and setting expiry at the end of
parsing.

=cut
package XML::SAX::Cache::Tee;

require 5.005;
use strict;
use warnings;

use base qw(XML::Filter::Tee);


sub new {
	my $proto = shift;
	my $class = ref($proto) || $proto;
	my ($cache_entry, $filter, @handlers) = @_;

	my $self = $class->SUPER::new(@handlers);
	bless $self, $class;

	$self->{SAXCacheEntry} = $cache_entry;
	$self->{SAXCacheFilter} = $filter;
	$self->{SAXCacheHandlers} = \@handlers;

	return $self;
}

sub cache_handlers {
	my $self = shift;
	return @{$self->{SAXCacheHandlers}};
}

sub end_document {
	my $self = shift;

	my $result = $self->SUPER::end_document(@_);

	# Document has ended, retrieve and store the validity object and
	# expiry
	my $validity = $self->{SAXCacheFilter}->cache_validity();
	$self->{SAXCacheEntry}->set_validity($validity) if defined $validity;

	my $expiry = $self->{SAXCacheFilter}->cache_expiry();
	$self->{SAXCacheEntry}->set_expiry($expiry) if defined $expiry;

	return $result;
}


1;
__END__

=head1 SEE ALSO

XML::SAX::Cacheable

=head1 AUTHOR

Chris Leishman <chris@leishman.org>

=head1 COPYRIGHT

Copyright (C) 2006 Chris Leishman.  All Rights Reserved.

This module is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY KIND,
either expressed or implied. This program is free software; you can
redistribute or modify it under the same terms as Perl itself.

$Id: Tee.pm,v 1.7 2006/02/04 14:38:18 caleishm Exp $

=cut


lib/XML/SAX/Cache/Entry.pm
=head1 NAME

XML::SAX::Cache::Entry - Interface for a cache store for XML::SAX::Cache

=head1 SYNOPSIS

  package XML::SAX::Cache::MyCache::Entry;
  use XML::SAX::Cache::Entry;

  our @ISA = qw(XML::SAX::Cache::Entry);

=head1 DESCRIPTION

This module defines the required interface for a cache entry for use with
XML::SAX::Cacheable.

Currently this interface is fully implemented by the Cache::Entry module (from
the Cache package on CPAN) which may be used as a concrete caching
implementation.

=head1 METHODS

=over

=cut
package XML::SAX::Cache::Entry;

require 5.005;
use strict;
use warnings;
use Storable qw(freeze thaw);

our $VERSION = '1.00';

=item $e->handle( $mode )

Returns an IO::Handle by which data can be read, or written, to the cache.
If there is no data in the cache, it should return undef.

=cut

sub handle;

=item my $val = $e->validity()

Returns the validity object that was stored with this cache entry.

=cut

sub validity;

=item $e->set_validity($val)

Store a validity object with this cache entry.

=cut

sub set_validity;

=item my $date = $e->expires()

Returns the expiry date (in unix time) for this cache entry.

=cut

sub expires;

=item $e->set_expiry($date)

Set the date (in unix time) that this cache entry should be automatically
expired.

=cut

sub set_expiry;


1;
__END__

=back

=head1 SEE ALSO

XML::SAX::Cacheable

=head1 AUTHOR

Chris Leishman <chris@leishman.org>

=head1 COPYRIGHT

Copyright (C) 2006 Chris Leishman.  All Rights Reserved.

This module is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY KIND,
either expressed or implied. This program is free software; you can
redistribute or modify it under the same terms as Perl itself.

$Id: Entry.pm,v 1.10 2006/02/04 14:38:18 caleishm Exp $

=cut


lib/XML/SAX/Cacheable.pm
=head1 NAME

XML::SAX::Cacheable - XML::SAX::Base interface extension for caching

=head1 SYNOPSIS

  package XML::Filter::MyFilter;
  use XML::SAX::Base;
  use XML::SAX::Cacheable;

  our @ISA = qw(XML::SAX::Base XML::SAX::Cacheable);

=head1 DESCRIPTION

This module extends the interface for XML::SAX to include methods for
controlling the caching of output.  It's default behaviour is to allow the
output of the SAX handler to be cached indefinitely.  This module requires no
initialization.

You can use this interface by simply adding it to your ISA inheritance list,
and then defining any methods you want to.

=head1 REQUIREMENTS

In order to add XML::SAX::Cacheable support using the default methods, your
module must support the get_handler() and set_handler() methods.  These are
provided by XML::SAX::Base automatically so modules using that as a base will
work fine.

=head1 METHODS

A number of methods are defined as the public interface for the caching
system.  Most methods have a default implementation.

=over

=cut
package XML::SAX::Cacheable;

require 5.005;
use strict;
use warnings;
use Carp;

our $VERSION = '1.00';
our @ISA = qw( Exporter );
our @EXPORT_OK = qw( make_cacheable );

=item $x->cacheable()

Returns 1 if the filter output is cacheable, otherwise returns undef.  In the
latter case users should NOT try to call any of the other cache_* methods on
the object, as it may not implement them.

The default implementation returns 1.

=cut

sub cacheable {
    return 1;
}

=item $x->cache_key()

If the output is not going to be consistent for the same input, but is likely
to occur regularly, the filter can return a true value from $x->cacheable(),
and return a key here that can be used to identify the specific variant it
will be outputting on the next run.

For example, a filter than inserts one of 5 random quotes into a document
could return a key string in the set [1..5] to indicate the specific variant.

Users of this interface should use this key when selecting a cache entry
object to supply to the other cache_* methods.  Note that there is no
requirement that the key returned here is globally unique - two filters
returning the same key here would still require different cache entry objects.

The default implementation returns undef (no key).

=cut

sub cache_key {
    return undef;
}

=item $x->cache_immutable($cache_entry)

Returns a boolean value (1 or 0) indicating whether the filter is immutable,
or undef if it is currently unknown.  A filter is immutable if it will always
result in the same output (given the same input) and thus has no dependencies
on other data sources.

When this method is called on the first run (before anything is cached) some
filters wont yet know if they have dependencies or not.  In this situation the
filter should return undef here and then at the end of the run store the
dependencies into the cache (using $cache_entry->set_validity($blah)) or at
least store an indication that there were no dependencies.  On later runs the
cache_entry should be queried (using $blah = $cache_entry->validity()) and the
dependency information extracted - thus allowing a defined result to be
returned.

Example:

    sub cache_immutable {
        my ($self, $cache_entry) = @_;
        # If there is no validity, then this is the first run
        my $validity = $cache_entry->validity();
        defined $validity or return undef;
        # If validity is empty then there are no dependencies
        return $validity? 1 : 0;
    }

The default implementation of this method returns 0.

=cut

sub cache_immutable {
    return 0;
}

=item $x->cache_check($cache_entry)

Returns a boolean value indicating whether the cache is up-to-date with
respect to any dependencies.  If necessary the module can determine this by
comparing with the validity object stored in the cache.

The default implementation returns 1, indicating the cache is always valid.

=cut

sub cache_check {
    return 1;
}

=item $x->cache_enable($cache_entry, [ $validity_only ] )

This method is invoked to direct the filter to cache it's output to the
$cache_entry.  This method should not be called a second time without calling
$x->cache_disable() first.

An optional second argument, $validity_only, is a boolean used to indicate
that the filter should only store a validity object (and expiry) into the
cache - and should not actually store any data.  This is useful if the user
wants to be able to determine when a filters output becomes inconsistent
with previous runs (by calling cache_check) but doesn't actually want to be
able to replay the data from the cache.  This may occur, for instance, with
SAX generators that will not benefit from caching - but where it is required
to know if their output is consistent in order to control caching further down
a SAX pipeline.  It is not required that the filter implement this behaviour,
it may ignore this flag and always cache, but it is strongly desirable.

The default implementation sets up caching by replacing the handler for the
filter with XML::SAX::Cache::Tee, that duplicates the output to the original
handler(s) and also to a consuming filter that stores the SAX events.  By
default that filter is a instance of XML::Filter::Recorder initialized using a
filehandle (which is obtained via $cache_entry->handle()), but this can be
changed by redefining the sax_recorder() method.

At the end of processing $x->cache_expiry() and $x->cache_validity() are
called to retrieve a expiry object and a validity object, which are then
stored in the cache using $cache_entry->set_expiry() and
$cache_entry->set_validity() respectively.

=cut

my @sax_handlers = qw(
    Handler ContentHandler DocumentHandler DTDHandler
    LexicalHandler ErrorHandler DeclHandler EntityResolver
);

sub cache_enable {
    my $self = shift;
    my ($cache_entry, $validity_only) = @_;

    # We require modules here here instead of 'using' above as subclasses
    # might override this method and nolonger need these classes loaded.

    # Subclass of XML::Filter::Tee that takes care of storing validity
    require XML::SAX::Cache::Tee;

    # Extract old handlers
    my %orig_handlers;
    foreach (@sax_handlers) {
        my $handler = $self->get_handler($_)
            or next;
        $orig_handlers{$_} = $handler;
    }

    if (my $handler = $orig_handlers{Handler}) {
        $handler->isa('XML::SAX::Cache::Tee')
                and croak "Cache has already been enabled";
    }

    # The existing handlers are stored into the specially crafted Tee
    # class, since we can't store anything into self as we don't know
    # what it's actual type is.
    my @tee_handlers = ( \%orig_handlers );

    # Add the recorder to tee_handlers - unless we're only caching the validity
    unless ($validity_only) {
        my $fh = $cache_entry->handle('>');
        my $cache_consumer = $self->sax_recorder($fh);
        push(@tee_handlers, $cache_consumer);
    }

    my $tee = XML::SAX::Cache::Tee->new($cache_entry, $self, @tee_handlers);
    
    # Set the default handler to go via the Tee filter, and remove the rest
    foreach (keys %orig_handlers) {
        $self->set_handler($_, undef);
    }
    $self->set_handler($tee);
}

=item $x->cache_disable()

This method is invoked to disable caching.

=cut

sub cache_disable {
    my $self = shift;

    my $tee = $self->get_handler();
    
    $tee->isa('XML::SAX::Cache::Tee')
        or croak "Cache has not been enabled";

    my @handlers = $tee->cache_handlers();
    my $orig_handlers = shift(@handlers);

    # Restore the handlers
    while (my ($key, $handler) = each %$orig_handlers) {
        $self->set_handler($key, $handler);
    }
}

=item $x->cache_playback($cache_entry)

This method causes the cached data to be 'played back' to the downstream SAX
handlers.  This method should NOT be called if caching is enabled (via
cache_enable).  In scalar context it returns a boolean result indicating
whether playback was attempted or whether the cache was empty.  In list
context it also returns the result from the recorder playback function as the
2nd element - this is typically whatever was returned by end_document in the
final handler.  Any failures during playback will be delivered as exceptions.

The user should call cache_check before attempting this method to ensure that
the result is up-to-date.  Then the return value of this method should be
checked to ensure the cached data hasn't expired.

=cut

sub cache_playback {
    my $self = shift;
    my ($cache_entry) = @_;

    if (my $handler = $self->get_handler()) {
        $handler->isa('XML::SAX::Cache::Tee')
            and croak "Can't playback when cache has been enabled";
    }

    my $fh = $cache_entry->handle('<');

    # return if stream is empty
    return undef if not $fh or eof($fh);

    my $cache_recorder = $self->sax_recorder($fh);

    foreach (@sax_handlers) {
        my $handler = $self->get_handler($_)
            or next;
        $cache_recorder->set_handler($_, $handler);
    }

    my $result = $cache_recorder->playback();

    return wantarray? ( 1, $result ) : 1;
}


=back

=head1 'PROTECTED' METHODS

These methods do not form part of the XML::SAX::Cacheable interface, and are
not required to be implemented by classes that support cacheability.  However
the default implementation of cache_check and cache_enable invoke these
'protected' methods to provide a mechanism for derived classes to control the
caching process.

=over

=item $x->cache_validity()

Returns a validity string (or serialisable object) for storing along with the
cached version.

=cut

sub cache_validity {
    return undef;
}

=item $x->cache_expiry()

Returns an integer value indicating the unix time that the cache should expire
at.

=cut

sub cache_expiry {
    return undef;
}

=item $x->sax_recorder($handle)

Returns a recorder object for storing SAX events to the cache (and playing
them back).  Returns an object of type XML::Filter::Recorder by default.

=cut

sub sax_recorder {
	my $self = shift;
	require XML::Filter::Recorder;
	return XML::Filter::Recorder->new(@_);
}


=back

=head1 MAKING A SAX FILTER CACHEABLE

Hopefully, over time, authors of SAX handlers will start to add the required 
cache-ability methods into their interface.  However, to add cache-ability to
any SAX handlers interface you can import the 'make_cacheable' function and
use it to add default implementations of all the cacheability methods into the
handlers interface.  It will only add the methods if the filters package
doesn't already have a 'cacheable' method.

Note that the default implementations mean the handlers output will be
considered indefinitely cacheable and there will be no dependencies
considered.

Use 'make_cacheable' as follows:

  use XML::SAX::Cacheable qw(make_cacheable);

  make_cacheable('XML::Filter::SomeFilter');

This will import the methods into the XML::Filter::SomeFilter package.
make_cacheable returns a true value (1) if it is successful, or a false value
(undef) otherwise.

=cut

sub make_cacheable {
    my $proto = shift;
    my $class = ref($proto) || $proto;

    return undef if $class->can('cacheable');

    my $code = "package $class;\n";

    foreach my $sub (
        'cacheable','cache_key', 'cache_immutable',
        'cache_check', 'cache_enable', 'cache_disable',
        'cache_playback', 'cache_validity', 'cache_expiry')
    {
        $code .= "*$sub = \\\&XML::SAX::Cacheable::$sub;\n";
    }

    $code .= "1;";
    eval $code or die $@;
    return 1;
}

1;
__END__

=head1 SEE ALSO

XML::SAX::Cache::Entry, XML::Filter::Recorder, XML::SAX::Base

=head1 AUTHOR

Chris Leishman <chris@leishman.org>

=head1 COPYRIGHT

Copyright (C) 2006 Chris Leishman.  All Rights Reserved.

This module is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY KIND,
either expressed or implied. This program is free software; you can
redistribute or modify it under the same terms as Perl itself.

$Id: Cacheable.pm,v 1.16 2006/02/06 06:11:58 caleishm Exp $

=cut
